### PR TITLE
Drag and drop ability on viewer

### PR DESCRIPTION
**Pull request by parveshneedhoo**
Right now, on the 3D-viewer measurement, we can place two spheres to measure the distance between two points. In this PR, we will add the ability to drag the sphere once it is placed.

### CODE REVIEW

File: src/measurements.ts [hunk 1]
```diff
@@ -12,41 +17,59 @@ const METERS_TO_CM = 100;
...
+
+    this.control.addEventListener('dragend', () => {
+      setTimeout(() => {
+        loader.unlockRotationAndClick();
+      }, 500);
+    });
+
+    this.control.addEventListener('drag', (event: any) => {
+      this.onSphereDrag(event.object as THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>);
+    });
```
Comment (zfir) [inline src/measurements.ts]: Better to handle this directly on loader and this will also reduce the number of parmeters passed on the constructor

**Review by ombr — COMMENTED** [review]
- [x] Table should not be displayed when loading
- [x] We should use a # instead of a query param ?
- [x] when we drag we should stay on the surface
- [ ] Nice to have but may be done a another pull-request (Bonus) save the lines added to the url ?

**Review by zfir — COMMENTED** [review]
We should remove all spheres when clearing all - when adding the first and not the second one, it does not clear all spheres

**Review by ombr — COMMENTED** [review]
Test url is not working ?

**Comment by parveshneedhoo** [comment]
> Test url is not working ?

Oops no @ombr 

I forgot to replace the "?" by "#" one the url.

It works now

However for the 3D file, I think it has been moved, I need to check with @zfir  to change the value in the environment variable

File: src/measurements.ts [hunk 2]
```diff
@@ -12,41 +17,70 @@ const METERS_TO_CM = 100;
...
+  ) {
     this.scene = scene;
+    this.group = group;
+    this.camera = camera;
 
-    this.measurementsTable = document.querySelector("#measurements-table-container") as HTMLTableElement;
-    if (this.measurementsTable) {
-      this.measurementsTable.style.display = "none";
-    }
+    this.control = new DragControls([], camera, renderer.domElement);
```
Comment (zfir) [inline src/measurements.ts]: Can we move this to a new class?

All the things about controls

File: src/measurements.ts [hunk 3]
```diff
@@ -156,29 +201,85 @@ export class Measurements {
...
+    this.spherePairs.forEach(pair => {
+      this.scene.remove(pair.sphere1);
+      this.scene.remove(pair.sphere2);
+      this.scene.remove(pair.line);
+      const objects = this.control.getObjects();
+      const index1 = objects.indexOf(pair.sphere1);
+      if (index1 > -1) objects.splice(index1, 1);
+      const index2 = objects.indexOf(pair.sphere2);
+      if (index2 > -1) objects.splice(index2, 1);
     });
```
Comment (zfir) [inline src/measurements.ts]:
```
this.spherePairs.forEach(({ sphere1, sphere2, line }) => {
  [sphere1, sphere2, line].forEach(object => this.scene.remove(object));
  this.control.getObjects().filter(obj => obj !== sphere1 && obj !== sphere2);
});
```

Does this work?

File: src/measurements.ts [hunk 4]
```diff
@@ -156,29 +201,85 @@ export class Measurements {
...
 
   private removeLine(index: number) {
-    this.scene.remove(this.lines[index]);
-    this.spheres[index].sphere1 && this.scene.remove(this.spheres[index].sphere1);
-    this.spheres[index].sphere2 && this.scene.remove(this.spheres[index].sphere2);
+    const pair = this.spherePairs[index];
+    if (pair) {
+      this.scene.remove(pair.line);
+      this.scene.remove(pair.sphere1);
+      this.scene.remove(pair.sphere2);
```
Comment (zfir) [inline src/measurements.ts:212]:
```suggestion
        [pair.line, pair.sphere1, pair.sphere2].forEach(obj => this.scene.remove(obj));
```

same here

File: src/index.ts [hunk 5]
```diff
@@ -3,8 +3,10 @@ import "./templates/styles.css";
 import { Utils } from "./utils";
 
 function getUrlParameter(name: string): string | null {
-  const urlParams = new URLSearchParams(window.location.search);
-  return urlParams.get(name);
+  const hash = window.location.hash.substring(1);
+  const hashParams = new URLSearchParams(hash);
+  const value = hashParams.get(name);
+  return value;
 }
```
Comment (zfir) [inline src/index.ts]: function getUrlParameter(name: string): string | null {
  return new URLSearchParams(window.location.hash.slice(1)).get(name);
}

Does this work?

**Review by zfir — COMMENTED** [review]
@parveshneedhoo Can we check if we can move the controls to another class

File: src/measurements.ts (hunk 3 above)
Comment (parveshneedhoo) [reply src/measurements.ts]: No

`this.control.getObjects().filter(obj => obj !== sphere1 && obj !== sphere2);` does not work.
It creates a filtered copy of getObjects() excluding sphere1 and sphere2, but it doesn’t update getObjects() directly.  And we can't even reassign it.

File: src/templates/index.html [hunk 6]
```diff
@@ -28,11 +28,11 @@
         <tbody id="measurements"></tbody>
       </table>
     </div>
-    <div id="loading">
+    <div id="loading" style="display: none;">
```
Comment (HashirRajah) [inline src/templates/index.html:31]: Why use inline styling?

File: src/drag-control-handler.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import * as THREE from "three";
+import { DragControls } from 'three/examples/jsm/controls/DragControls';
+import { Loader } from "./loader";
+import { Measurements } from "./measurements";
+
+export class DragControlHandler {
+  private control: DragControls;
```
Comment (HashirRajah) [inline src/drag-control-handler.ts:7]:
```suggestion
  private controls: DragControls;
```

**Review by HashirRajah — APPROVED** [review]

**Review by ombr — APPROVED** [review]
Behavior is good for me, I did not check all the code.

File: src/drag-control-handler.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  private onMouseDown: (mouseEvent: MouseEvent) => void;
+  private onMouseMove: (mouseEvent: MouseEvent) => void;
+  private isMouseMoveListenerActive: boolean = false;
+
+  constructor(
+    camera: THREE.Camera,
+    renderer: THREE.WebGLRenderer,
+    private loader: Loader,
```
Comment (zfir) [inline src/drag-control-handler.ts:15]:
```suggestion
   loader: Loader,
```

File: src/drag-control-handler.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    private onDrag: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>) => void,
+    private onDragEnd: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>, event: any) => void,
```
Comment (zfir) [inline src/drag-control-handler.ts:17]:
```suggestion
   onDragEnd: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>, event: any) => void,
```

File: src/drag-control-handler.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  public removeObjectFromControl(spheres: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>[]) {
+    spheres.forEach((sphere) => {
+      const objects = this.control.getObjects();
+      const index = objects.indexOf(sphere);
+      if (index > -1) objects.splice(index, 1);
+    });
+  }
+
+  private handleDragStart(event: any, measurement: Measurements) {
```
Comment (zfir) [inline src/drag-control-handler.ts]:
```suggestion
  private handleDragStart(event: Event, measurement: Measurements) {
```

Does this work?

File: src/drag-control-handler.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        this.isMouseMoveListenerActive = false;
+      }
+    };
+    window.addEventListener('mouseup', onMouseUp);
+    setTimeout(() => {
+      this.loader.unlockRotationAndClick();
+    }, 500);
+  }
+
+  private handleOnDrag(event: any) {
```
Comment (zfir) [inline src/drag-control-handler.ts]:
```suggestion
  private handleOnDrag(event: Event) {
```

File: src/drag-control-handler.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    private onDrag: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>) => void,
+    private onDragEnd: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>, event: any) => void,
+    measurements: Measurements
+  ) {
+    this.control = new DragControls([], camera, renderer.domElement);
+    this.onMouseDown = (mouseEvent: MouseEvent) => measurements.updateMouse(mouseEvent);
+    this.onMouseMove = (mouseEvent: MouseEvent) => measurements.updateMouse(mouseEvent);
+
+    this.control.addEventListener('dragstart', (event: any) => this.handleDragStart(event, measurements));
+    this.control.addEventListener('drag', (event: any) => this.handleOnDrag(event));
```
Comment (zfir) [inline src/drag-control-handler.ts]:
```suggestion
    this.control.addEventListener('drag', this.handleOnDrag);
```

File: src/drag-control-handler.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    this.control.addEventListener('dragend', (event: any) => this.handleDragEnd(event, measurements));
```
Comment (zfir) [inline src/drag-control-handler.ts]:
```suggestion
    this.control.addEventListener('dragend', (event: Event) => this.handleDragEnd(event, measurements));
```

Check if this works

File: src/measurements.ts [hunk 14]
```diff
@@ -12,88 +17,101 @@ const METERS_TO_CM = 100;
...
+    const intersects = this.raycaster.intersectObject(this.group, true);
     if (intersects.length > 0) {
-      this.addPoint(intersects[0].point);
-      if (this.points.length === 2) {
-        this.drawLine();
-        this.calculateDistance();
-        this.points = [];
-        this.generatedColor = undefined;
-      }
+      this.addPoint(intersects[0].point, this.generatedColor);
```
Comment (zfir) [inline src/measurements.ts:64]:
```suggestion
      this.addPoint(intersects[0].point);
```

File: src/measurements.ts [hunk 15]
```diff
@@ -12,88 +17,101 @@ const METERS_TO_CM = 100;
...
       (event.clientX / window.innerWidth) * 2 - 1,
       -(event.clientY / window.innerHeight) * 2 + 1
     );
   }
 
-  private addPoint(point: THREE.Vector3) {
-    this.points.push(point);
-    const sphere = this.createSphere(point);
-    this.spheresForLine.push(sphere);
+  private addPoint(point: THREE.Vector3, color: string) {
```
Comment (zfir) [inline src/measurements.ts:75]:
```suggestion
  private addPoint(point: THREE.Vector3) {
```

File: src/measurements.ts [hunk 16]
```diff
@@ -12,88 +17,101 @@ const METERS_TO_CM = 100;
...
+    if (this.currentSpheres.length === 2) {
+      const [sphere1, sphere2] = this.currentSpheres;
+      const pair: SpherePair = {
+        sphere1,
+        sphere2,
+        line: new THREE.Line(
+          new THREE.BufferGeometry().setFromPoints([sphere1.position, sphere2.position]),
+          new THREE.LineBasicMaterial({ color })
+        ),
+        color,
```
Comment (zfir) [inline src/measurements.ts:89]:
```suggestion
        this.generatedColor,
```

File: src/measurements.ts [hunk 17]
```diff
@@ -140,45 +167,145 @@ export class Measurements {
     }).join("");
 
     measurementsContainer.innerHTML = rows;
-    this.addEventListeners();
+    
+    this.addEventListenersOnRemoveLineButton();
+    if (!this.listenersAdded) {
```
Comment (zfir) [inline src/measurements.ts]:
```suggestion
    if (!this.eventListenersAdded) {
```

or eventListenersInitialised

File: src/measurements.ts [hunk 18]
```diff
@@ -140,45 +167,145 @@ export class Measurements {
...
+  }
+
+  private onSphereDrag(draggedSphere: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>) {
+    const pair = this.spherePairs.find(p => p.sphere1 === draggedSphere || p.sphere2 === draggedSphere);
+    if (!pair) return;
+
+    this.raycaster.setFromCamera(this.mouse, this.camera);
+    const intersects = this.raycaster.intersectObject(this.group, true);
+    if (intersects.length > 0) {
+      this.lastValidPosition = intersects[0].point;
```
Comment (zfir) [inline src/measurements.ts]:
```suggestion
      this.lastIntersectPoint = intersects[0].point;
```

File: src/loader.ts [hunk 19]
```diff
@@ -54,11 +56,45 @@ export class Loader {
 
     container.appendChild(this.renderer.domElement);
     
-    this.measurement = new Measurements(this.scene);
+    this.measurement = new Measurements(this.scene, this.camera, this.renderer, this.group, this);
     window.addEventListener("resize", this.onWindowResize.bind(this), false);
     window.addEventListener("beforeunload", this.cleanup.bind(this));
     window.addEventListener("unload", this.cleanup.bind(this));
-    container.addEventListener("click", this.onMouseClick.bind(this), false);
+    container.addEventListener("mousedown", this.onMouseDown.bind(this), false);
```
Comment (zfir) [inline src/loader.ts:63]: Not sure if better to use Arrow funcctions here. Can be done in another PR

File: src/drag-control-handler.ts (hunk 8 above)
Comment (parveshneedhoo) [reply src/drag-control-handler.ts:15]: https://www.amitmerchant.com/auto-implemented-properties-in-typescript/

File: src/drag-control-handler.ts (hunk 9 above)
Comment (zfir) [inline src/drag-control-handler.ts:17]:
```suggestion
    private onDragEnd: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>, event: Event) => void,
```

File: src/drag-control-handler.ts [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+  private onMouseMove: (mouseEvent: MouseEvent) => void;
+  private isMouseMoveListenerActive: boolean = false;
+
+  constructor(
+    camera: THREE.Camera,
+    renderer: THREE.WebGLRenderer,
+    private loader: Loader,
+    private onDrag: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>) => void,
+    private onDragEnd: (object: THREE.Mesh<THREE.SphereGeometry, THREE.MeshBasicMaterial>, event: any) => void,
+    measurements: Measurements
```
Comment (zfir) [inline src/drag-control-handler.ts:18]:
```suggestion
    measurements: Measurements
```

Move above privates

File: src/drag-control-handler.ts [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+      const index = objects.indexOf(sphere);
+      if (index > -1) objects.splice(index, 1);
+    });
+  }
+
+  private handleDragStart(event: THREE.Event, measurement: Measurements) {
+    window.addEventListener('mousedown', this.onMouseDown);
+    this.loader.lockRotationAndClick();
+		
+    measurement.lastIntersectedPosition = event.object.position.clone();
```
Comment (zfir) [inline src/drag-control-handler.ts:45]: use callback in measurement

File: src/drag-control-handler.ts [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  private handleDragEnd(event: THREE.Event, measurement: Measurements) {
+    const onMouseUp = (mouseEvent: MouseEvent) => {
+      measurement.updateMouse(mouseEvent);
```
Comment (zfir) [inline src/drag-control-handler.ts:50]:
```suggestion
      measurement.updateMouse(mouseEvent);
```
use measurement in measurement

File: src/loader.ts [hunk 23]
```diff
@@ -54,11 +56,45 @@ export class Loader {
 
     container.appendChild(this.renderer.domElement);
     
-    this.measurement = new Measurements(this.scene);
+    this.measurement = new Measurements(this.scene, this.camera, this.renderer, this.group, this);
```
Comment (zfir) [inline src/loader.ts:59]:
```suggestion
    this.measurement = new Measurements(this.scene, this.camera, this.renderer, this.group);
```

Handle all events in drag-control-handler

File: src/measurements.ts [hunk 24]
```diff
@@ -12,88 +17,101 @@ const METERS_TO_CM = 100;
...
-      this.spheres.push({ sphere1: this.spheresForLine[0], sphere2: this.spheresForLine[1] });
-      this.spheresForLine = [];
+    if (this.currentSpheres.length === 2) {
+      const [sphere1, sphere2] = this.currentSpheres;
+      const pair: SpherePair = {
+        sphere1,
+        sphere2,
+        line: new THREE.Line(
+          new THREE.BufferGeometry().setFromPoints([sphere1.position, sphere2.position]),
+          new THREE.LineBasicMaterial({ color })
```
Comment (zfir) [inline src/measurements.ts:87]:
```suggestion
          new THREE.LineBasicMaterial({ color:  this.generatedColor })
```

File: src/measurements.ts [hunk 25]
```diff
@@ -111,24 +129,33 @@ export class Measurements {
...
+    
+    if (this.spherePairs.length === 0 && measurementsTableContainer) {
+      measurementsTableContainer.style.display = "none";
+      this.currentSpheres.forEach(sphere => {
+        this.scene.remove(sphere);
+        this.dragControlHandler.removeObjectFromControl([sphere]);
+      })
+      this.spherePairs = [];
+      this.currentSpheres = [];
+      this.generatedColor = undefined;
```
Comment (zfir) [inline src/measurements.ts:143]: Duplication should we move that to a new function

**Review by zfir — COMMENTED** [review]
- Should be handle all the events from loader and measurement in Drag Control Handler?
- Loader -> Measurement -> DragControlHandler : I think we handle the classes like and for example DragControlHandler should not have loader as a global variable

I will merge it like this. The comments can be handled in the next PR.

Reviewed with @parveshneedhoo

**Pull request merged**